### PR TITLE
[train][tests] Test xgboost and lightgbm data parallel training

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

Before this PR, `test_lightgbm_trainer` and `test_xgboost_trainer` didn't actually validate that we were doing data parallel training. This PR validates this by asserting model equality across workers, which it does by comparing validation loss instead of model parameters because the latter is more brittle (e.g. json metadata key comparisons). One thing to call out is that in `test_lightgbm_trainer`, I do not shard the validation dataset across workers to ensure all workers compute the same validation score (see https://github.com/microsoft/LightGBM/issues/4392), while in `test_xgboost_trainer`, I do shard the validation dataset across workers because xgboost [allreduces metrics](https://github.com/dmlc/xgboost/blob/d0135d0f43ff91e738edcbcea54e44b50d336adf/python-package/xgboost/callback.py#L131). I also documented all of the above in the unit tests themselves.

# Testing

Unit tests. I also manually set breakpoints in the above tests to verify that lightgbm does not allreduce metrics but xgboost does.

### CODE REVIEW

File: python/ray/train/v2/tests/test_xgboost_trainer.py [hunk 1]
```diff
@@ -78,6 +94,10 @@ def xgboost_train_fn_per_worker(
             num_boost_round=remaining_iters,
             xgb_model=starting_model,
         )
+        collector.report.remote(
+            ray.train.get_context().get_world_rank(),
+            evals_result["valid"]["rmse"],
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/train/v2/tests/test_xgboost_trainer.py]: fixed.

**Review by justinvyu — APPROVED** [review]
